### PR TITLE
Telemetry Handler Update.

### PR DESCRIPTION
**Pull request by peombwa**
This PR updates the telemetry handler spec to include the following host environment metadata to help use make informed decisions when it comes to framework version upgrades and native platform optimizations.
- HostOS.
- RuntimeEnvironment.

### CODE REVIEW

File: middleware/TelemetryHandler.md [hunk 1]
```diff
@@ -2,18 +2,38 @@
...
 
 ## Requirements
 
 - Add a `client-request-id` header with GUID value to request if one is not present.  If the [RequestContext](../middleware/RequestContext.md) has a `ClientRequestId` property set then that value should be used.
-- Add `SdkVersion` request header to each request to identify the language and version of the client SDK library.  The product identifier should be of the format `graph-{lang}`. The `SdkVersion` version can contain multiple values, or appear multiple times in a request. Values should be comma delimited as normal header lists are if the `SdkVersion` contains multiple values. e.g.
 
-    `SdkVersion: MyApp/1.0, spfx/1.0, graph-javascript-1.0`
+- Add `SdkVersion` request header to each request to identify the language and version of the client SDK library. The product identifier should be of the format `graph-{lang}`. The `SdkVersion` version can contain multiple values, or appear multiple times in a request. Values should be comma delimited as normal header lists are if the `SdkVersion` contains multiple values. e.g.
```
Comment (MIchaelMainer) [inline middleware/TelemetryHandler.md]: The SdkVersion MUST be updated with information identifying the use of not only the core libraries, but also the service API libraries.

File: middleware/TelemetryHandler.md [hunk 2]
```diff
@@ -2,18 +2,38 @@
...
+
+- Add `RuntimeEnvironment` request header to capture the runtime framework on which the client SDK is running on. Ideally, we should capture the runtime environment in the form of `RuntimeEnvironment: Name/Version`. e.g.
+
+    - `RuntimeEnvironment: .NETFramework/1.1` for .NET.
+    - `RuntimeEnvironment: Node/1.1` for JavaScript.
+    - `RuntimeEnvironment: JRE/1.1` for Java.
+
+#### Ideal Metadata Capture
+```cs
+SdkVersion: graph-dotnet/1.16.0 (featureUsage=0f)
```
Comment (MIchaelMainer) [inline middleware/TelemetryHandler.md]: `SdkVersion: graph-dotnet-core/1.16.0 (featureUsage=0f) graph-dotnet-beta/0.6.0-preview`

File: middleware/TelemetryHandler.md (hunk 1 above)
Comment (peombwa) [reply middleware/TelemetryHandler.md]: For .Net, we can capture the service models' SDK version in our core library via reflections, but I'm not 100% sure we can do the same for other client libraries unless we update our code generation pipeline to capture this.

For now, I'll update to the spec to  treat this as an optional metadata, and we should only capture it when it's available.

File: middleware/TelemetryHandler.md [hunk 3]
```diff
@@ -2,17 +2,47 @@
 
 ## Objectives
 
-Provide a middleware component that attaches metadata to a Graph request in order help improve the SDK team improve the developer experience.
+Provide a mandatory middleware component that attaches metadata to a Graph request in order help the SDK team improve the developer experience.
```
Comment (darrelmiller) [inline middleware/TelemetryHandler.md]: How are we making this mandatory?  Surely someone could just build a pipeline without it?  I know we would rather have devs include it, but I think mandatory is overstating it a bit.

File: middleware/TelemetryHandler.md [hunk 4]
```diff
@@ -2,17 +2,47 @@
...
+    - `RuntimeEnvironment: .NETFramework/1.1` for .NET.
+    - `RuntimeEnvironment: Node/1.1` for JavaScript.
+    - `RuntimeEnvironment: JRE/1.1` for Java.
+
+#### Ideal Metadata Capture
+```
+SdkVersion: graph-dotnet-beta/0.6.0-preview, graph-dotnet-core/1.16.0 (featureUsage=0f)
+client-request-id: fdae6861-5916-486d-93d9-9129160b2d79
+HostOS: Microsoft Windows 10.0.18362
+RuntimeEnvironment: .NETFramework/4.7.2
```
Comment (darrelmiller) [inline middleware/TelemetryHandler.md]: I'm still not convinced we should introduce two new headers.  
Why not? 

    SdkVersion: graph-dotnet-beta/0.6.0-preview, graph-dotnet-core/1.16.0 (featureUsage=0f),  .NETFramework/1.1, Windows/10.0.18362

Comment (pschaeflein) [reply middleware/TelemetryHandler.md]: Should the community project be adding to that header?

Comment (darrelmiller) [reply middleware/TelemetryHandler.md]: It certainly can.  Spfx, the Graph Toolkit and Postman collections do, so why not.

File: middleware/TelemetryHandler.md (hunk 3 above)
Comment (peombwa) [reply middleware/TelemetryHandler.md]: My initial thinking was to make it ubiquitous whenever a Http client is created using our client factory. We can do this by internally registering the `TelemetryHandler` as the last `DelegatingHandler` when creating a pipeline. It won't be configurable by devs, but we are the ones who will set it. Here is a [POC](https://github.com/microsoftgraph/msgraph-sdk-dotnet-core/blob/ad92e8ace29d5850a8f92593449da00b2a30adda/src/Microsoft.Graph.Core/Requests/GraphClientFactory.cs#L207) for this. I should rephrase it to say mandatory when created using our client factory.

File: middleware/TelemetryHandler.md (hunk 4 above)
Comment (peombwa) [reply middleware/TelemetryHandler.md]: Won't this abstract the environment values being put in the header i.e. hostOS and runtimeEnvironment. How about we introduce an extra header e.g. `HostEnvironment`, to hold the two values (hostOS and runtimeEnvironment).

On second thought, since the headers are meant to be used by us, we can control how we siphon info out of them, thus your suggestion is 100% valid and keeps everything contained in one header.
